Add the execve system call version 1.

takeharukato opened the pull request:
Hi,

As you said in Environment variable #15 , xv6 cannot handle environment variable properly.
One of the reasons of this, it does not have execve(2) system call.
I wrote a patch to implement execve(2) for xv6 as a trial.

Would you please look into it and give it a try if you can?

P.S. Implementing exec(2) system call in a user land library might be more preferable way.
In addition to this, I did not modify the shell application. Some modification might be needed for the shell. 

Thanks,
Regards,

takeharukato committed:
Add the execve system call version 1.

takeharukato committed:
Merge branch 'master' into devel-tkato-execfix
A confilict on kernel/defs.h was fixed.

takeharukato committed:
Fix conflicts with the upstream.

takeharukato commented:
Hi, 

Apparently, you or main stream authors might add shebang feature.
I've fixed major conflicts somehow but my patch does not care of this feature, I'll consider how we should treat it this weekend.

Of course,  you can drop this PR if this PR  puts your burden.
I just think that some code examples might be useful for your study.

Thanks,
Regards,

takeharukato committed:
Fix adding extra option for the compiler by mistake.

takeharukato committed:
Fix space differences.

tyfkda commented on ulib/crt0.c:
```diff
@@ -1,8 +1,38 @@
+  fprintf(stderr, "crt0: envp: %p environ: %p\n", envp, environ);
+#endif  /*  DEBUG_SHOW_EXECVE_ENVS  */
+
   int code = main(argc, argv);
```
I heard `main` function also accepts `envp` through third argument on Linux.

takeharukato replied:
The C language specification defines the prototype of main function and it says that the main function has two arguments, argc, argv. 
Strictly speaking, Linux's behavior which you said is not preferable so much in this meaning.
We should not depends on this behavior at least.
Would you please see http://www.dii.uchile.cl/~daespino/files/Iso_C_1999_definition.pdf?

FYI, on the other hand, some said that POSIX says main function has three arguments from runtime library point of view. From this idea, Linux just decided to conform to the POSIX standard,  however, as far fs I checked the execve definition in POSIX IEEE1003.1 standard System Interfaces, Issue 6,  it says main function has two arguments. But some believe this behavior is correct  at least.

Anyway, roughly speaking, some says that taking three arguments in main is better way from compatibility among OSes ( e.g., UNIX Internals: A Practical Approach https://www.amazon.com/dp/020187721X ) even though user programs do not permitted 
to invoke the main function with three args.

Which do you like?
I'll fix it as your favor.

Thanks,
Regards,

tyfkda commented on kernel/exec.c:
```diff
@@ -218,3 +221,161 @@ exec(const char *path, const char* const *argv)
 }
+
+int
+execve(const char *path, const char *argv[], const char *envp[]){
```
Could you replace `execelf` function to accept `envp` and use it in this function?
Also, could you remove `exec` function?

takeharukato replied:
It is done.
I've pushed it.

tyfkda commented on kernel/syscall.h:
```diff
@@ -25,3 +25,4 @@
 #define SYS_isatty 24
 #define SYS_ftruncate 25
 #define SYS_readdir 26
+#define SYS_execve  27
```
Could you remove `SYS_exec` and implement `exec` function in user library?

takeharukato replied:
As you said, it is needed to implement PATH parsing to get rid of exec syscall and to implement it in userland library.
Apparently, we've start writing environment variable handing facility in  shell program, 
would you please continue this work and implement exec function in ulib by yourself?
To be honest, I am not able to have an interesting about implement the exec function in libc.

Thanks,
Regards,

tyfkda reviewed (commented):
@takeharukato , I didn't know that environment variables are handled in user land.

I am thinking that adding `execve` syscall is cumbersome,
why don't you replace the `exec` syscall for the purpose?

takeharukato committed:
Re-writing execve(2), ELF loader, and shebang function as tyfkda's comment.

takeharukato commented:
I've removed all exec(2) invocation from user programs and merged your shebang function, my elf loader for execve, and execve syscall handler.
Then, I've pushed these fixes.
Would you please review them?
I show my simple test  result on my box for your information.
```
xv6...
acpi: cpu#0 apicid 0
acpi: cpu#1 apicid 1
acpi: ioapic#0 @fec00000 id=0 base=0
cpu1: starting 1
cpu0: starting 0
sb: size 1000 nblocks 942 ninodes 200 nlog 30 logstart 2 inodestart 32 bmap start 57
init: starting sh
$ test.sh
abc
(exitcode = 0)
$ ls
.                 1     1      512  1970/01/01 09:00
..                1     1      512  1970/01/01 09:00
bin               1     2      384  1970/01/01 09:00
console           3    25        0  2018/08/19 03:34
uart              3    26        0  2018/08/19 03:34
(exitcode = 0)
$
```
Thanks,
Regards,

takeharukato committed:
Remove exec system call from kernel.

takeharukato commented:
On second thought, it is more preferable to remove exec(2) in kernel from now on as you said.
And more, I did not remove exec call in usertest.
I fixed it and removed exec(2).

takeharukato committed:
Add environment variable handling.

takeharukato committed:
Merge https://github.com/tyfkda/xv6 into devel-tkato-execfix and resolve the conflict on user/sh.c. I also add -std=gnu99 to C compiler flags.

takeharukato commented:
Hi,

After all, I've written a patch to support environment variables in userland library( getenv, setenv ) and
implement an exec(2) in userland as you want.
Would you please check it?

P.S. find_path function in kernel/exec.c might be no longer needed.  I'll add this fix later.


Thanks, 
Regards,

takeharukato committed:
Remove find_path in kernel/exec.c.

takeharukato commented:
Hi,

FYI, I've pushed modifications to remove find_path.
Now, you can handle environment variables in userlands to your heart's content:-).

P.S. If my memory is correct, $? is NOT an environment variable but a kind of shell variables. You should not set it by setenv (or putenv).

Thanks,
Regards,

takeharukato commented:
Hi,

>  I didn't know that environment variables are handled in user land.

Sure.
Many people misunderstand the fact that OS kernel does not know anything about environment variables.  These variables are owned by shells.


P.S. I did not notice that shebang is handled by OS kernel in my young days:-).  
I was very surprised when I learned this. 

Anyway, have a fun!

Thanks,
Regards,

tyfkda commented on ulib/exec.c:
```diff
@@ -0,0 +1,192 @@
+			goto free_env_out;
+	}
+	envs[i] = 0;
+	dir = path_refer(i);
```
Prevent searching paths if the given path is absolute or relative (e.g. `./ls`, `/bin/ls`),

takeharukato replied:
Right.
I've fixed it.

takeharukato replied:
FYI, I forgot writing that strictly speaking, it is more preferable from security point of view to ensure the first element of path is a directory when it invoke first execve(2) systemcall. 
But I think that this shell is not mature enough to consider this issue.

tyfkda commented on user/test.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
```
There is `clr` command written in shell script,
you can use it for test
and you don't have to add extra file.

takeharukato replied:
Sure.
I've removed my test.

tyfkda commented on ulib/ulib.c:
```diff
@@ -2,6 +2,8 @@
 #include "sys/stat.h"
 #include "unistd.h"
 
+extern char **environ;
```
It looks `environ` is not used in this file.

takeharukato replied:
Right.
I've removed it.

tyfkda commented on Makefile:
```diff
@@ -3,7 +3,8 @@
 BITS = 64
 XOBJS = obj/knl/vm64.o
-XFLAGS = -m64 -DX64 -mcmodel=kernel -mtls-direct-seg-refs -mno-red-zone
+STD_FLAGS = -std=gnu99
```
If adding compile option doesn't related to the change,
please remove it to minimize changes.

takeharukato replied:
You've already  use C99 extensions in your additional codes so strictly speaking, it is needed by your modification. Original xv6 does not use C99 syntax but your code require this option.
Did you notice that?
Some distribution adopt newer gcc so they accept C99 as default, so you might not notice this.
Of course, I can remove -std=gnu99 option,  I also think that C99 option issue should be isolated from this topic.
Anyway, would you please consider whether we should remove it or not again?

tyfkda replied:
Please remove it.

tyfkda commented on include/stdlib.h:
```diff
@@ -14,3 +14,7 @@ void free(void*);
+
+char *getenv(const char *);
+int setenv(const char *, const char *, int );
+
```
Could you also declare `unsetenv`?

takeharukato replied:
Sure.
I've added it.

tyfkda commented on kernel/sysfile.c:
```diff
@@ -442,12 +444,57 @@ sys_exec(void)
+}
+
+int
+sys_execve(void)
```
Now `sys_exec` is not used anymore,
could you replace it to `sys_execve`?

tyfkda commented on kernel/initcode64.S:
```diff
@@ -18,12 +19,26 @@ exit:
+  .string "TERM=xv6\0"
+
+init_env_path:
+  .string "PATH=/sbin:/bin:/usr/sbin:/usr/bin\0"
```
Could you change the value to be `/bin` only,
since the system doesn't have other directories?

takeharukato replied:
As I wrote in a comment bellow, this means the default PATH for application.
So it is important to ensure security as far as we can even though directories do not exist in practice.
Because the file system layout is determined by users, I think that  we should do is trying to keep it safe whenever we can.

tyfkda replied:
Please use `/bin` only.

tyfkda commented on kernel/exec.c:
```diff
@@ -6,61 +6,55 @@
-static struct inode *find_path(const char *path, char *resultPath, int bufsiz) {
-  struct inode *ip;
+#define EXECVE_USTK_ARGS_NR       (4)
+#define EXECVE_SHEBANG_RESV_ARGS  (2)
```
It looks `EXECVE_SHEBANG_RESV_ARGS` is not used.

takeharukato replied:
Sure.
I've removed it.

takeharukato commented:
FYI,

As I wrote in comments, the following codes are desirable but some timing problem might exist in other portion in the kernel code. Apparently, it seems that this code increases the probability of occurrence of the issue. I'll look into the cause of this issue as long as I can spare time for it.
If you have a interesting in solving this problem, would you apply following diff.

```
diff --git a/kernel/exec.c b/kernel/exec.c
index c8b1694..7950f95 100644
--- a/kernel/exec.c
+++ b/kernel/exec.c
@@ -238,10 +238,12 @@ load_shebang(const char *path, const char *argv[], const char *envp[])
 #endif  /*  DEBUG_SHOW_SHEBANG  */

   /*
-   * Note: Invoking execve is desirable
-   * but it seems kernel memory is not enough at this moment.
+   * Invoking execve to handle user program again.
+   * It is needed to invoke shebang using shell command
+   * like #!/bin/script.sh
+   * Note that script.sh might contain shebang(#!).
    */
-  return load_elf(argv2[0], argv2, envp);
+  return execve(argv2[0], argv2, envp);
```
How to reproduce the issue is easy, you just invoke test.sh shebang repeatedly. 
```
$ make qemu
xv6...
acpi: cpu#0 apicid 0
acpi: cpu#1 apicid 1
acpi: ioapic#0 @fec00000 id=0 base=0
cpu1: starting 1
cpu0: starting 0
sb: size 1000 nblocks 942 ninodes 200 nlog 30 logstart 2 inodestart 32 bmap start 57
init: starting sh
$ test.sh
abc
(exitcode = 0)
$ test.sh
**pid 10 sh: trap 14 err 4 on cpu 0 eip 0x1334 addr 0x18000--kill proc
(exitcode = 190)**

tyfkda commented on ulib/exec.c:
```diff
@@ -0,0 +1,192 @@
+	}
+	printf("Exec:%s %p %p\n", path, argv, envs);
+#endif  /*  SHOW_EXECV_PARAM  */
+	rc = execve(path, argv, envs);
```
Prevent searching from current path
if the given path is not absolute nor relative.

takeharukato replied:
I re-pushed the fix for this comment because I mishandled path/a.out case.

tyfkda commented on kernel/exec.c:
```diff
@@ -6,61 +6,55 @@
-  }
-  return ip;
-}
+//#define DEBUG_SHOW_EXECVE_ENVS
```
Please remove debug labels and code blocks.

takeharukato replied:
Sure.
I've removed it.

tyfkda commented on ulib/crt0.c:
```diff
@@ -1,8 +1,65 @@
 
-void _start(int argc, char** argv) {
+static void
+setup_environment(char *envs[]){
```
Could you move this function into environ.c?

takeharukato replied:
Right,
But would you please notice that if we move it into  environ.c, setup_environment cause a name space pollution.
I've changed its function name to _setup_environment to minimize  pollution.
Anyway, some external function might remain at any case.

tyfkda commented on ulib/crt0.c:
```diff
@@ -1,8 +1,65 @@
+
+#define PATH_ENV_VAR "PATH"
+
+char **environ={0};
```
Could you move this variable into environ.c?

takeharukato replied:
Sure.
I've fixed it.

takeharukato committed:
Fix for shebang with shebang.

tyfkda commented on kernel/exec.c:
```diff
@@ -6,61 +6,55 @@
+//#define DEBUG_SHOW_EXECVE_ENVS
+//#define DEBUG_SHOW_SHEBANG
 
-// Check ELF header
```
Would it be possible to minimize the change,
just adding environment handling
and keep function names?

takeharukato replied:
After all, I tried to keep functions as it is as far as I can. Would you please believe it.
But to make shebang work correctly,  it will be needed to call execve after shebang processes are completed. To avoid dead locks in such situation, we should lock inode when we need it indeed.
I thought that I had no choice but rewriting codes which you wrote. Sorry.

tyfkda commented on kernel/exec.c:
```diff
@@ -122,99 +158,144 @@ execelf(const char *progname, const char *path, const char* const *argv,
 }
 
-int
-execshebang(const char *path, const char * const *argv,
```
Would it be possible to keep this function as is,
except passing environment?

takeharukato replied:
Introducing environment variables changes the logic of exec ( strictly speaking, exec in xv6 should be called execv in POSIX). So I can keep function name only.

takeharukato replied:
FYI, in addition to this, there is a race condition between checking a binary type and execution.
To avoid this race, we should lock inode for path from execve  to execelf or check binary type for each binary handler. 
However, as far as I know,  POSIX does not require this because this might cause a tremendous performance issue when we meet cascaded shebang cases.
Of course I can make execve lock the inode between execve and execelf and execshbang.
But I am not sure which is more preferable. At this moment, xv6 does not have asynchronous I/O facility, inode bitmap, and anonymous block I/O to improve inode and block bitmap handling. So this lock might not cause serious performance problem at least.
However if you will continue to improve xv6, it might be an important design decision.

What do you think?

Anyway, I'll check POSIX specification again.

P.S. I might have used namei in the beginning of execve and puti it at the end of execve to prevent removing a file. But I'm not sure it is what you want and it is needed from the standard.

takeharukato replied:
On second thought, I think that execve should be written as follows at least to keep the compatibility to your codes:
```
-    if ( rc != 0 )
-      goto error_out;
-  } else if ( ( rd_size >= EXECVE_SHEBANG_STRLEN ) &&
+    if ( rc == 0 )
+      goto out;
+  }
+
+  if ( ( rd_size >= EXECVE_SHEBANG_STRLEN ) &&
     ( memcmp((void *)&exeblk[0], EXECVE_SHEBANG_STR , 2) == 0 ) ) {

     rc = execshebang(path, argv,envp);
-    if ( rc != 0 )
-      goto error_out;
+    if ( rc == 0 )
+      goto out;
   }
-  else
-    panic("execve: unknwon binary");

+  panic("execve: unknwon binary");
+
```
But I'm not sure ilock handling in your codes were written intentionally  or not.
I’m sorry to bother you, would you please notice that again, if you want lock inode between exec and execelf and execshebang.

takeharukato replied:
Hi,

After all, I found the intention of your code.
In my understanding, You would like to make much more account to consumption of kernel stack than inode locking. Is it correct?

I think that it can be achieved by checking correctness of files in each binary handlers like Linux kernel does. Certainly, Linux kernel pass an opened file descriptor to each binary handler, passing an opened  inode to execelf and execshebang might be good  in a sense to avoid a race condition precisely. But I think that inode lock cause problems easily so I did not write such codes yet. 

BTW, I found the cause of the crash which I mentioned above ( **pid 10 sh: trap 14 err 4 on cpu 0 eip 0x1334 addr 0x18000--kill proc ). This is caused by shortage of kernel stack. I found your code is correct in this meaning. Sorry to my misunderstanding. 

I show the kernel stack dump when I met the problem.
In the following case, for example, the kernel stack was allocated at the page which was started from 0xffffffff8df4f000 page but it pointed at the address in the page which was started from 0xffffffff8df4e000 .
```
0xffffffff8df4efd0:     0x0000000000000000      0x0101010101010101 <= kstack points 0xffffffff8df4efd0 but it should have been in 0xffffffff8df4f000 page.
0xffffffff8df4efe0:     0x0000000000000008      0x0000000000000202
0xffffffff8df4eff0:     0xffffffff8df4f000      0x0000000000000000
0xffffffff8df4f000:     0xffffffff8df5c098      0x0000000000014000
0xffffffff8df4f010:     0xffffffff8df4f050      0xffffffff8010984e
0xffffffff8df4f020:     0xffffffff8df4f030      0x0000000000000000
0xffffffff8df4f030:     0xffffffff8df5e000      0x0000000000007f78
0xffffffff8df4f040:     0xffffffff80116fe8      0xffffffff8df4f710
0xffffffff8df4f050:     0xffffffff8df4f070      0xffffffff80109c1d
0xffffffff8df4f060:     0xffffffff8df5e000      0xffffffff8df4fdb0
0xffffffff8df4f070:     0xffffffff8df4f670      0xffffffff80101c3e<=execve ( 0xffffffff80101c3e is placed in exec function ) frame in 0xffffffff8df4f000 page.
0xffffffff8df4f080:     0x0101010101010101      0xffffffff8df4f430
0xffffffff8df4f090:     0x0000000000000002      0x0000000000007f98
0xffffffff8df4f0a0:     0x0000000000007fb0      0x0000000000000003
0xffffffff8df4f0b0:     0xffffffff8df4f410      0x0000000000008000
0xffffffff8df4f0c0:     0xffffffff80116fe8      0xffffffff8df42000
```

Anyway, I've fixed the problem and pushed it.  Sorry.
But it get stability again.

P.S. BTW,  `char            line[512];`  in execshebang might be too big, 
I did not touch it to keep your codes. PATH_MAX which has proper length for the system might be more suitable for this. Strictly speaking, we should have create limits.h in include/ and put the macro into it. I defined and used PATH_MAX in ulib/exec.c because no one use it for now. You can move it into include/limits.h but note that limits.h is one of freestanding headers. You should define this macro carefully not to conflict with PATH_MAX defined by C compiler when you add it.

Thanks,
Regards,

takeharukato replied:
I've consulted with POSIX ( SUSv4 ), there is no description about mutual exclusion of file access when exec family functions perform. So I think both your code and mine might be correct. 

BTW, new POSIX specification (2017) defines arguments in main function more precisely like:
```
Some implementations provide a third argument to main() called envp. This is defined as a pointer to the environment. The ISO C standard specifies invoking main() with two arguments, so implementations must support applications written this way. Since this volume of POSIX.1-2017 defines the global variable environ, which is also provided by historical implementations and can be used anywhere that envp could be used, there is no functional need for the envp argument. Applications should use the getenv() function rather than accessing the environment directly via either envp or environ. Implementations are required to support the two-argument calling sequence, but this does not prohibit an implementation from supporting envp as an optional third argument.
```
According to this definition, we should treat the environment array as the third argument of main function from _start, but the C prototype declaration for main function should have two argument only ( As prestigious man in UNIX, Mr. Pate said ).

tyfkda replied:
I think it can be possible to preserve original functions and code blocks
and just add `envp` handling.

If you modify other things at the same time,
it is difficult to see the point of this PR
and difficult to find out introducing bugs accidentally.

tyfkda commented on kernel/initcode64.S:
```diff
@@ -18,12 +19,26 @@ exit:
+# char *envp[] = { init_env, 0 };
+.p2align 2
+envp:
+  .quad init_env_term
```
Could you remove `TERM` variable
since the system doesn't have any `TERM` handling?

takeharukato replied:
Humm, I thought  you tried to write a line discipline once.
I thought that TERM variable is needed to write  a line discipline and terminal handling correctly. 

OK, anyway, I've removed it.

BTW, Initial PATH should have been secure path ( /sbin:/bin:/usr/sbin:/usr/bin ) ideally.
What do you think?
P.S. Note that the existence of directories are not concerned with initial PATH.

tyfkda reviewed (commented):
@takeharukato , first of all, thank you for your effort!

I'm afraid putting my preference,
but would you follow the indentation rules to use two spaces to be consistent,
and minimize code changes?

takeharukato committed:
Some fixes from tyfkda's comment.

takeharukato committed:
Fix path handling when user spcify absolute or dot relative path.

takeharukato commented:
Hi,

I re-identified my codes by 2 spaces as far as I can.
I'm sorry in advance if I forgot fixing something.

P.S. It remind me of the fact that xv6 comes from MIT:-).

Thanks,
Regards,

takeharukato committed:
Fix wrong handling of relative path and absolute path( e.g.,  ./path/a.out, /path/a.out and path/a.out ).
I mishandled path/a.out case.

takeharukato committed:
Remove extra file checks.

takeharukato committed:
Remove unused macros ( I forgot removing them before ).

tyfkda commented on ulib/exec.c:
```diff
@@ -0,0 +1,161 @@
+#include "fcntl.h"
+#include "file_def.h"
```
"fcntl.h" and "file_def.h" are not required.

tyfkda commented on ulib/exec.c:
```diff
@@ -0,0 +1,161 @@
+  if ( PATH_MAX <= len ) 
+    len = PATH_MAX - 1;
+  s[len] = '\0';
+  if ( (len - 1) != '\0' ) {
```
`len > 1` ?

tyfkda commented on ulib/exec.c:
```diff
@@ -0,0 +1,161 @@
+   * First, we try to execute a binary with relative path and absolute path
+   * ( e.g.,  ./path/a.out, /path/a.out and path/a.out ).
+   */
+  execve(path, argv, envs);
```
Prevent searching from current directory
when path doesn't include separator(`/`).
(e.g. typing `ls` doesn't execute `./ls`)

tyfkda commented on ulib/exec.c:
```diff
@@ -0,0 +1,161 @@
+  /*
+   * Second, we try to execute a binary according to PATH environment variable.
+   */
+  dir = path_refer(i);
```
Prevent searching from `PATH`s
when path includes separator.

tyfkda commented on kernel/exec.c:
```diff
@@ -69,16 +57,16 @@ execelf(const char *progname, const char *path, const char* const *argv,
-      goto bad;
     if(ph.vaddr % PGSIZE != 0)
       goto bad;
+    if((sz = allocuvm(pgdir, sz, PGROUNDUP(ph.vaddr + ph.memsz))) == 0)
```
Why do you change the order,
and use `PGROUNDUP`?
Please keep as is in this PR.

tyfkda commented on user/init.c:
```diff
@@ -28,7 +30,7 @@ runsh(int dev, const char* devname)
     }
     if(pid == 0){
-      exec("sh", argv);
+      execve("/bin/sh", argv, environ);
```
Because you modified `exec` library function to keep same behavior,
it is not necessary to change init.c at all.

tyfkda commented on ulib/environ.c:
```diff
@@ -0,0 +1,250 @@
+#define ENV_VAR_LEN   (128)
+#define ENV_VAR_DELIM   '='
+
+char **environ={0};
```
Global variable `environ` is not necessary
if you revert init.c.

tyfkda commented on kernel/exec.c:
```diff
@@ -88,32 +76,62 @@ execelf(const char *progname, const char *path, const char* const *argv,
+     * and we ensure each element is aligned to the word.
+     */
+    sp = (sp - (strlen(envp[nr_envs - env_idx - 1]) + 1)) & ~(sizeof(uintp)-1);
+    if (copyout(pgdir, sp, envp[nr_envs - env_idx - 1], 
```
Is there any reason to scan `envp` in reverse order?
I think scanning in normal order makes code mush simpler,
and it can be more easily comparable to the original code for `argv`.

tyfkda commented on ulib/environ.c:
```diff
@@ -0,0 +1,250 @@
+
+    p = &evar_array[i];
+		
+    if ( p->var != 0 ) {
```
Can you use `NULL` instead of `0`?

takeharukato replied:
Sorry, I have a question.
Can we use NULL in kernel codes?
Apparently, xv6 does not use NULL in kernel codes because it is designed to provide a "flavor" of Unix version 6. So it does not use NULL. I would also like to use NULL also and it is preferable to improve code maintainability.

tyfkda replied:
Files under ulib directory are user library,
so I think it's acceptable to use `NULL`.

tyfkda commented on ulib/exec.c:
```diff
@@ -0,0 +1,161 @@
+  char *name, *next;
+  
+  new_path = strdup(path_val);
+  if ( new_path == 0 ) {
```
Can you use `NULL`?

takeharukato committed:
Fix according to tyfkda's comments.

takeharukato committed:
Resolve a conflict in include/unistd.h.

takeharukato commented:
Hi,

After all, I've fixed my codes as you like.
I hope that I can understand what you want correctly.
But, finally, find_path in kernel/exec.c could not be removed even though if you've wanted to remove it as you said before. I can remove exec function in kernel/exec.c only.
I could not solve this request, sorry.

Thanks,
Regards,

takeharukato committed:
Use NULL macro in userland.

takeharukato committed:
Make behavior of exec POSIXly correct.

tyfkda commented on ulib/exec.c:
```diff
@@ -0,0 +1,153 @@
+     * First, we try to execute a binary according to PATH environment variable
+     * when path does not contain any slash.
+     */
+    dir = path_refer(i);
```
`i` is not initialized to `0`,
so the value is changed according to the number of environment variables.

tyfkda commented on ulib/exec.c:
```diff
@@ -0,0 +1,153 @@
+   * Second, we try to execute a binary with relative path and absolute path
+   * ( e.g.,  ./path/a.out, /path/a.out and a.out/ ).
+   */
+  rc = execve(path, argv, envs);
```
Please do not fall through the above `if` statement
to prevent searching from the current path:
https://github.com/tyfkda/xv6/pull/25/files/bf283c95ea6d885f42d212a0eb8f6c9f5f5f086b#r211915309

Please enclose this statement in `else` clause.

tyfkda commented on kernel/initcode64.S:
```diff
@@ -22,8 +23,19 @@ exit:
 init:
   .string "init\0"
```
File path is not searched from `PATH` in `execve` syscall,
the path should be `/bin/init`.

tyfkda commented on kernel/exec.c:
```diff
@@ -205,9 +224,9 @@ exec(const char *path, const char* const *argv)
+    result = execelf(path, path, argv, envp, &elf, &ip, &pgdir);
   } else {
-    result = execshebang(exepath, argv, &ip, &pgdir);
+    result = execshebang(exepath, argv, envp, &ip, &pgdir);
```
Would you replace `exepath` to `path`
(and remove `exepath` local variable)
when you removed `find_path`?

tyfkda reviewed (commented):
Thank you for the changes.

Could you use `namei` function to replace `find_path`,
and remove it (and `FILE_SEPARATOR`) in kernel/exec.c?

takeharukato committed:
Fix according to comment.

takeharukato committed:
Fix exec path handling.

takeharukato commented:
(1) changes 56854bc
     I regarded your request is removing find_path if other parts would be changed.

(2) changes e1da032
    Finally, I could notice what you said and you are right.
   Certainly, it's my fault. Sorry.　I've  overlooked `*Otherwise*, the path prefix for this file is obtained by a search of the directories passed as the environment variable PATH.` and `If this environment variable is not present, the results of the search are implementation-defined.` in POSIX specification. So I thought I needed to do this to keep backward compatibility with xv6's exec. And more, if I should have done that thing, I should search the execution file from the current directory only if PATH environment variable is not defined and if path does not contain \<slash\>.

Thanks,
Regards,

takeharukato commented:
FYI, if you want to implement putenv(3), you need to change the environment variable data structure from the list to an array.

I wrote environment variable handling in userland as same way as your original environment variable routines (except I used double linked list instead of single linked list) .
As far as I know, some old runtimes implements adopt this way.
 I thought it is not so wrong because we do not provide putenv(3), and this way fulfill the requirements of  getenv(3) and setenv(3). 
Thus, I adopt your way, but strictly speaking, it will break putenv specification when you add putenv(3).

takeharukato commented:
FYI, speaking of NULL, xv6 uses zero page for userland programs, so NULL should NOT be zero.

Original xv6 is bankrupt in this meaning.

Another wrong NULL handling is that NULL is defined in unistd.h, it should be defined in stddef.h in the manner of ISO/ANSI C.

Thanks,
Regards,

takeharukato commented:
FYI,
After all, strictly speaking, this PR does not implement execve, it just introduce an environment variable argument into XV6's exec and emulate current YOUR exec system call behavior into userland.
This means this execve is not implemented in the standard conformance point of view.

I write remained issues around exec family and environment variable handing in your codes for your hacking.

(1) Exec in kernel does not handle memory segment which contains .bss section when the binary is strange.
     =>It just cause killing a caller process. If execve has a responsible to ensure to clear .bss section in xv6 
          specific ELF binary, the kernel should round memsz up. But the current behavior might also be 
          permitted by POSIX.   This is an answer of issue #19. 
(2) Execve shall also handle PATH environment variable from POSIX conformance point of view.
(3) Shebang facility should handle executables which uses shebang with shell script (e.g., #!/bin/script.sh).
(4) Get rid of inode locks between execve and exec handlers. They should be holding inode only.
(5) Environment variables shall export as environ global variable, it is required by ISO/ANSI-C.
(6) NULL shall be defined in stddef.h it is required by ISO/ANSI-C.
(7) Environment variables should be handled by an array instead of lists. It is virtually  required by ISO/ANSI-C.
(8) Exec should handle the error condition properly when the basename of path is directory.
(9) Introduce ETXTBSY, 
    When processes open the file with write flag, the kernel mark it into inode.
    And then, execve check this mark to determine whether some processes open the file to write.
    If a writer process exists, execve shall return an error with ETXTBSY. It is the right solution of your ilock ussage in exec.
(10) The system shell should be fixed to handle command invocation procedure POSIXly correct.
        The exec family is not responsible for POSIX's "Command Search and Execution
" behavior. For example, `kill` command shall be invoked if a command path name does not contain any \<slash\> characters and even though PATH environment does not contain the directory where the kill command exists.


Thanks,
Regards,

tyfkda commented:
@takeharukato , thank you for your effort.
I've merged your changes
(with squashing commits and some modification).
